server/src/main/java/org/elasticsearch/index/mapper/flattened/FlattenedFieldParser.java

```diff
@@ -32,44 +36,64 @@ class FlattenedFieldParser {
     static final String SEPARATOR = "\0";
     private static final byte SEPARATOR_BYTE = '\0';
 
-    private final String rootFieldName;
-    private final String keyedFieldName;
+    private final String rootFieldFullPath;
+    private final String keyedFieldFullPath;
 
     private final MappedFieldType fieldType;
     private final int depthLimit;
     private final int ignoreAbove;
     private final String nullValue;
 
     FlattenedFieldParser(
-        String rootFieldName,
-        String keyedFieldName,
+        String rootFieldFullPath,
+        String keyedFieldFullPath,
         MappedFieldType fieldType,
         int depthLimit,
         int ignoreAbove,
         String nullValue
     ) {
-        this.rootFieldName = rootFieldName;
-        this.keyedFieldName = keyedFieldName;
+        this.rootFieldFullPath = rootFieldFullPath;
+        this.keyedFieldFullPath = keyedFieldFullPath;
         this.fieldType = fieldType;
         this.depthLimit = depthLimit;
         this.ignoreAbove = ignoreAbove;
         this.nullValue = nullValue;
     }
 
-    public List<IndexableField> parse(final DocumentParserContext context) throws IOException {
-        XContentParser parser = context.parser();
+    public List<IndexableField> parse(final DocumentParserContext documentParserContext) throws IOException {
+        XContentParser parser = documentParserContext.parser();
         XContentParserUtils.ensureExpectedToken(XContentParser.Token.START_OBJECT, parser.currentToken(), parser);
 
+        XContentBuilder rawDataForSyntheticSource = null;
+        if (documentParserContext.mappingLookup().isSourceSynthetic() && ignoreAbove < Integer.MAX_VALUE) {
+            var copyingParser = new CopyingXContentParser(parser);
+            rawDataForSyntheticSource = copyingParser.getBuilder();
+            parser = copyingParser;
+        }
+
         ContentPath path = new ContentPath();
         List<IndexableField> fields = new ArrayList<>();
 
+        var context = new Context(parser, documentParserContext);
         parseObject(context, path, fields);
+        if (rawDataForSyntheticSource != null && context.isIgnoredValueEncountered()) {
+            // One or more inner fields are ignored due to `ignore_above`.
+            // Because of that we will store whole object as is in order to generate synthetic source.
```

Comment:
Let's create an issue to track improving this which we can prioritise later.

Reply:
#110239